src/crimson/thread/ThreadPool.h

```diff
@@ -0,0 +1,95 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*- 
+// vim: ts=8 sw=2 smarttab
+
+#include <atomic>
+#include <condition_variable>
+#include <tuple>
+#include <type_traits>
+#include <boost/lockfree/queue.hpp>
+#include <boost/optional.hpp>
+#include <core/future.hh>
+#include <core/semaphore.hh>
+
+#include "Condition.h"
+
+namespace ceph::thread {
+
+struct WorkItem {
+  virtual ~WorkItem() {}
+  virtual void process() = 0;
+};
+
+template<typename Func, typename T = std::invoke_result_t<Func>>
+struct Task final : WorkItem {
+  Func func;
+  boost::optional<T> result;
+  ceph::thread::Condition on_done;
+  seastar::semaphore& free_slot;
+public:
+  Task(seastar::semaphore& fs, Func&& f)
+    : func(std::move(f)),
+      free_slot(fs)
+  {}
+  void process() override {
+    result = func();
+    on_done.notify();
+  }
+  seastar::future<T> get_future() {
+    return on_done.wait().then([this] {
+        free_slot.signal();
+	auto res = std::move(result.get());
+	delete this;
```

Comment:
i'm guessing that `seastar::engine().read_some()` in `wait()` can throw (return an exceptional future), which would leak this task. the call to `free_slot.wait()` in `submit()` could as well

so either Task should be ref-counted, or both cases should free the task on exceptions using `then_wrapped()` or `handle_exception()`